Create.c

```diff
@@ -91,6 +97,63 @@ int default_layout(struct supertype *st, int level, int verbose)
 	return layout;
 }
 
+static int discard_device(const char *devname, unsigned long long size)
+{
+	uint64_t range[2] = {0, size};
+	unsigned long buf[4096 / sizeof(unsigned long)];
+	unsigned long i;
+	int fd;
+
+	fd = open(devname, O_RDWR | O_EXCL);
+	if (fd < 0) {
+		pr_err("could not open device for discard: %s\n", devname);
+		return 1;
+	}
+
+	if (ioctl(fd, BLKDISCARD, &range)) {
+		pr_err("discard failed on '%s': %m\n", devname);
+		goto out_err;
+	}
+
+	if (read(fd, buf, sizeof(buf)) != sizeof(buf)) {
+		pr_err("failed to readback '%s' after discard: %m\n", devname);
+		goto out_err;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(buf); i++) {
+		if (buf[i]) {
+			pr_err("device did not read back zeros after discard on '%s': %lx\n",
+			       devname, buf[i]);
+			goto out_err;
+		}
+	}
+
+	close(fd);
+	return 0;
+
+out_err:
+	close(fd);
+	return 1;
+}
+
+static int discard_devices(struct context *c, struct mddev_dev *devlist,
+			   unsigned long long size)
+{
+	struct mddev_dev *dv;
+
+	for (dv = devlist; dv; dv = dv->next) {
+		if (!strcmp(dv->devname, "missing"))
+			continue;
+
+		if (c->verbose)
+			pr_err("discarding all data on: %s\n", dv->devname);
+		if (discard_device(dv->devname, size))
```

Comment:
how long is this operation? should we parallelize it?

Reply:
Instant. Parallelizing it is way to much complexity for the purpose.